File: 12_bermudan_swaption_lsm_pricing/python/lsm_path_simulation.py
"""
lsm_path_simulation.py

Monte Carlo path simulation for Bermudan swaption LSM pricing.

This chapter reuses the Chapter 09 HullWhiteModel object.
"""

import numpy as np


def simulate_hull_white_paths_for_lsm(
    model,
    r0,
    maturity,
    n_steps,
    n_paths,
    seed=42
):
    """
    Simulate Hull-White short-rate paths for LSM.

    Parameters
    ----------
    model : HullWhiteModel
        Chapter 09 Hull-White model.

    r0 : float
        Initial short rate.

    maturity : float
        Final simulation horizon.

    n_steps : int
        Number of simulation steps.

    n_paths : int
        Number of Monte Carlo paths.

    seed : int
        Random seed.

    Returns
    -------
    times : np.ndarray
        Simulation time grid.

    rate_paths : np.ndarray
        Short-rate paths with shape:

            (n_paths, n_steps + 1)
    """

    np.random.seed(seed)

    dt = maturity / n_steps

    times = np.linspace(
        0.0,
        maturity,
        n_steps + 1
    )

    rate_paths = np.zeros(
        (
            n_paths,
            n_steps + 1
        )
    )

    rate_paths[:, 0] = r0

    for step in range(n_steps):

        t = times[step]

        z = np.random.normal(
            size=n_paths
        )

        r = rate_paths[:, step]

        drift = np.array(
            [
                model.short_rate_drift(
                    t,
                    ri
                )
                for ri in r
            ]
        )

        rate_paths[:, step + 1] = (
            r
            +
            drift * dt
            +
            model.sigma
            *
            np.sqrt(dt)
            *
            z
        )

    return times, rate_paths


def path_discount_factors_to_index(
    times,
    rate_paths,
    end_index
):
    """
    Pathwise discount factors from time 0 to a time index.

    D(0,t_i) = exp(- integral_0^t_i r_s ds)
    """

    dt = times[1] - times[0]

    integrals = np.sum(
        rate_paths[:, :end_index],
        axis=1
    ) * dt

    return np.exp(
        -integrals
    )


def path_discount_factors_between_indices(
    times,
    rate_paths,
    start_index,
    end_index
):
    """
    Pathwise discount factors between two time indices.

    D(t_i,t_j) = exp(- integral_t_i^t_j r_s ds)
    """

    dt = times[1] - times[0]

    integrals = np.sum(
        rate_paths[:, start_index:end_index],
        axis=1
    ) * dt

    return np.exp(
        -integrals
    )


def nearest_time_index(
    times,
    target_time
):
    """
    Return nearest simulation-grid index.
    """

    return int(
        np.argmin(
            np.abs(
                times - target_time
            )
        )
    )

File: 12_bermudan_swaption_lsm_pricing/cpp/LSMPathSimulation.cpp
#include "LSMPathSimulation.hpp"

#include <random>
#include <cmath>
#include <algorithm>

/*
    Simulate Hull-White short-rate paths:

        r_{t+dt}
        =
        r_t
        +
        [theta(t) - a r_t] dt
        +
        sigma sqrt(dt) Z
*/

LSMPathSimulationResult simulateHullWhitePathsForLSM(
    const HullWhiteModel& model,
    double r0,
    double maturity,
    int nSteps,
    int nPaths,
    unsigned int seed
){

    double dt =
        maturity
        /
        static_cast<double>(
            nSteps
        );

    LSMPathSimulationResult result;

    result.times.resize(
        nSteps + 1
    );

    for(int i = 0; i <= nSteps; ++i){

        result.times[i] =
            i
            *
            dt;
    }

    result.ratePaths =
        std::vector<std::vector<double>>(
            nPaths,
            std::vector<double>(
                nSteps + 1,
                0.0
            )
        );

    for(int p = 0; p < nPaths; ++p){

        result.ratePaths[p][0] =
            r0;
    }

    std::mt19937 generator(seed);

    std::normal_distribution<double> normal(
        0.0,
        1.0
    );

    for(int step = 0; step < nSteps; ++step){

        double t =
            result.times[step];

        for(int path = 0; path < nPaths; ++path){

            double r =
                result.ratePaths[path][step];

            double z =
                normal(generator);

            double drift =
                model.shortRateDrift(
                    t,
                    r
                );

            result.ratePaths[path][step + 1] =
                r
                +
                drift
                *
                dt
                +
                model.sigma
                *
                std::sqrt(dt)
                *
                z;
        }
    }

    return result;
}

/*
    Nearest time index.
*/

int nearestTimeIndex(
    const std::vector<double>& times,
    double targetTime
){

    int bestIndex =
        0;

    double bestDiff =
        std::abs(
            times[0]
            -
            targetTime
        );

    for(size_t i = 1; i < times.size(); ++i){

        double diff =
            std::abs(
                times[i]
                -
                targetTime
            );

        if(diff < bestDiff){

            bestDiff = diff;
            bestIndex = static_cast<int>(i);
        }
    }

    return bestIndex;
}

/*
    Discount factor from time 0 to index.
*/

double pathDiscountFactorToIndex(
    const std::vector<double>& times,
    const std::vector<double>& ratePath,
    int endIndex
){

    double dt =
        times[1]
        -
        times[0];

    double integral =
        0.0;

    for(int i = 0; i < endIndex; ++i){

        integral +=
            ratePath[i]
            *
            dt;
    }

    return
        std::exp(
            -integral
        );
}

/*
    Discount factor between two time indices.
*/

double pathDiscountFactorBetweenIndices(
    const std::vector<double>& times,
    const std::vector<double>& ratePath,
    int startIndex,
    int endIndex
){

    double dt =
        times[1]
        -
        times[0];

    double integral =
        0.0;

    for(int i = startIndex; i < endIndex; ++i){

        integral +=
            ratePath[i]
            *
            dt;
    }

    return
        std::exp(
            -integral
        );
}
